working/0508 - strict-inference/strict-inference.md

```diff
@@ -0,0 +1,362 @@
+
+```dart
+fn(List<int> numbers) => print(numbers.first.isEven);
+
+void main() {
+  var args = ["one", "two", "three"];
+  var useArgs = true;
+  fn(useArgs ? args : []);
+}
+```
+
+Depending on the value of `useArgs`, the last line will result in a runtime
+error. The developer thinks that `args` is an appropriate value to send to
+`fn`, because surely static analysis would have reported if this wasn't so. In
+real world cases of this code, `fn` and `main` may be authored by different
+developers, and may be located in separate packages. `args` may have been
+created in a third location, by a third developer, with a complex generic type.
+Static analysis is supposed to help developers catch where they have
+misunderstood an API or the type of an object they are handling.
+
+The developer might think that the type of `[]` is unimportant, or that it is
+inferred from the `numbers` parameter type, or that it is inferred from the
+expression to the left of `:`. None of these are true. The type of `[]` is
+inferred from the types of its elements. As their are none, it "falls back" to
+`dynamic`. Then type of the conditional expression `useArgs ? args : []` is
+inferred from the "then" and "else" expressions, `args` (`List<String>`) and
+`[]` (`List<dynamic>`), resulting in LUB of the two, `List<dynamic>`.
+
+Static analysis allows a `List<dynamic>` argument for a `List<int>`, as an
+implicit cast. At runtime, however, `args` (a `List<String>`) fails to cast to
+`List<int>`.
+
+To prevent such an error, an empty collection literal (`[]`, `{}`) without an
+explicit type argument, and whose type cannot be inferred from downwards
+inference is considered to have an inference failure. In this example, the
+strict inference failure would report that the type of `[]` cannot be inferred,
+and suggest that the developer add an explicit type argument. The developer
+would likely add `<String>`, thinking "`args` is a `List<String>` and I think
+that `fn` accepts a `List<String>`, so I'll make the empty list alternative
+also a `List<String>`." At this point, existing static analysis will inform the
+developer that a `List<String>` cannot be passed where a `List<int>` is
+expected.
+
+Let's look at a second example. Consider the signature of [`Iterable.fold`]:
+
+> T fold <T>(T initialValue, T combine(T previousValue, E element))
+
+```dart
+void main() {
+  var a = [1, 2, 3].fold(true, (s, x) => s + x);
+}
+```
+
+There are no compile-time, static analysis errors in this code, but it is
+guaranteed to produce a failure at runtime, when `true + 1` is executed. The
+issue here is similar to the previous one: the developer likely thinks that the
+type of `s` (`T`) will be inferred from the type of `initialValue`, and that
+static analysis would report any issue with that type. But inference doesn't
+flow between parameters like that.  Instead, while trying to infer `T`, there
+is not enough information from downwards inference (the type of `a`), and there
```

Comment:
This isn't quite an accurate description of what happens.  What happens is that the downwards context provides no information.  Upwards inference constrains `T` to be a super type of `bool` from the first argument.  For the second argument, inference assumes that `s` and `x` have type `dynamic`, and hence infers `dynamic Function(dynamic, int)` as the type of the function literal.  This in turn constraints `T` to be a supertype of `dynamic`.  The LUB of `bool` and `dynamic` is `dynamic`, and so that is what is chosen for `T`.